File: Year_2026/February/balance_bst.rb
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val = 0, left = nil, right = nil)
        @val = val
        @left = left
        @right = right
    end
end

def balance_bst(root)
    nodes = []
    fill(root , nodes)
    return build(nodes , 0 , nodes.length - 1)
end

def fill(root , nodes)
    return nil unless root
    fill(root.left , nodes)
    nodes.push(root)
    fill(root.right , nodes)
end

def build(nodes , left , right)
    return nil if left > right
    mid = left + ((right - left) >> 1)
    root = nodes[mid]
    root.left = build(nodes , left , mid - 1)
    root.right = build(nodes , mid + 1 , right)
    return root
end

File: Year_2026/February/balance_bst.java
package February;
import java.util.*;

public class balance_bst {
   public static void main(String[] args){
      
   }
   static class Solution {
    private List<TreeNode> nodes;
    public TreeNode balanceBST(TreeNode root) {
        this.nodes = new ArrayList<>();
        fill(root);
        return build(0 , nodes.size() - 1);
    }
    private void fill(TreeNode root){
        if(root == null) return;
        fill(root.left);
        nodes.add(root);
        fill(root.right);
    }
    private TreeNode build(int start , int end){
        if(start > end) return null;
        int mid = start + ((end - start) >> 1);
        TreeNode root = nodes.get(mid);
        root.left = build(start , mid - 1);
        root.right = build(mid + 1 , end);
        return root;
    }
}
}
